editor/src/components/canvas/ui-jsx-canvas-renderer/scene-root.tsx

```diff
@@ -96,10 +107,23 @@ function useGetValidTemplatePaths(
   topLevelElementName: string | null,
   scenePath: ScenePath,
 ): Array<InstancePath> {
-  const topLevelElements = useContextSelector(RerenderUtopiaContext, (c) => c.topLevelElements)
+  const topLevelJSXComponentsMap: Map<string, UtopiaJSXComponent> = useEditorState((store) => {
```

Comment:
there is a problem with memoization here: since this is a Map, it will not pass shallow equal, because we create a new Map inside every time. I think there's not much point in keeping this a Map, so we should turn this into a MapLike object instead. that would fix the shallow equality, and the breaking render count test

Reply:
actually, it turns out shallowEquals did not at all handle Maps correctly. since that is a potentially huge problem, we decided to fix it for Maps. Now this shallowEquals works fine for this maps case here too, so disregard my previous comment!